balancer/pickfirst_leaf/pickfirst_leaf.go

```diff
@@ -0,0 +1,626 @@
+	}
+	curAddr := b.addressIndex.currentAddress()
+	sd, ok := b.subConns.Get(curAddr)
+	if !ok {
+		var err error
+		// We want to assign the new scData to sd from the outer scope, hence
+		// we can't use := below.
+		sd, err = newSCData(b, curAddr)
+		if err != nil {
+			// This should never happen, unless the clientConn is being shut
+			// down.
+			b.logger.Warningf("Failed to create a subConn for address %v: %v", curAddr.String(), err)
+			// The LB policy remains in TRANSIENT_FAILURE until a new resolver
+			// update is received.
+			b.state = connectivity.TransientFailure
+			b.addressIndex.reset()
+			b.cc.UpdateState(balancer.State{
+				ConnectivityState: connectivity.TransientFailure,
+				Picker:            &picker{err: fmt.Errorf("failed to create a new subConn: %v", err)},
+			})
+			return
+		}
+		b.subConns.Set(curAddr, sd)
+	}
+
+	scd := sd.(*scData)
+	switch scd.state {
+	case connectivity.Idle:
+		scd.subConn.Connect()
+	case connectivity.TransientFailure:
+		if !b.addressIndex.increment() {
+			b.endFirstPass(scd.lastErr)
+			return
+		}
+		b.requestConnection()
+	case connectivity.Ready:
+		// Should never happen.
+		b.logger.Errorf("Requesting a connection even though we have a READY subconn")
+	}
+}
+
+// updateSubConnState handles subConn state updates.
+// Only executed in the context of a serializer callback.
+func (b *pickfirstBalancer) updateSubConnState(sd *scData, state balancer.SubConnState) {
+	// Previously relevant subconns can still callback with state updates.
+	// To prevent pickers from returning these obsolete subconns, this logic
+	// is included to check if the current list of active subconns includes this
+	// subconn.
+	if activeSD, found := b.subConns.Get(sd.addr); !found || activeSD != sd {
+		return
+	}
+	if state.ConnectivityState == connectivity.Shutdown {
+		return
+	}
+
+	if state.ConnectivityState == connectivity.Ready {
+		b.shutdownRemaining(sd)
+		b.addressIndex.seekTo(sd.addr)
+		b.state = connectivity.Ready
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Ready,
+			Picker:            &picker{result: balancer.PickResult{SubConn: sd.subConn}},
+		})
+		return
+	}
+
+	// If we are transitioning from READY to IDLE, reset index and re-connect when
+	// prompted.
+	if b.state == connectivity.Ready && state.ConnectivityState == connectivity.Idle {
+		// Once a transport fails, we enter idle and start from the first address
+		// when the picker is used.
+		b.state = connectivity.Idle
+		b.addressIndex.reset()
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Idle,
+			Picker:            &idlePicker{exitIdle: b.ExitIdle},
+		})
+		return
+	}
+
+	if b.firstPass {
+		switch state.ConnectivityState {
+		case connectivity.Connecting:
+			// We can be in either IDLE, CONNECTING or TRANSIENT_FAILURE.
+			// If we're in TRANSIENT_FAILURE, we stay in TRANSIENT_FAILURE until
+			// we're READY. See A62.
+			// If we're already in CONNECTING, no update is needed.
```

Comment:
The `we` here is confusing because the connectivity state applies to the whole LB policy as well as to the subchannel. So, maybe be explicit about what the `we` is. Thanks.